docs/spec_draft.md

```diff
@@ -343,6 +344,66 @@ logical operation.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.batch_norm_inference
+
+### Semantics
+
+Normalizes the `operand` tensor across batch and spatial dimensions, for each
+feature in the `feature_index` dimension and produces a `result` tensor. The
+algorithm for `Batch Norm Inference` is same as 
+[Batch Norm Training](#stablehlobatch_norm_training), except it uses `mean`
+and `variance` provided in the input instead of computing them.
```

Comment:
I took some time to compare: 1) the algorithm in the paper, 2) the previous notation that used MLIR, 3) the current formulation - and think what would be the best for the spec.

Re: 1)/3), the algorithm in the spec is algebraically the same as what's used by [BatchNormExpander](https://github.com/tensorflow/tensorflow/blob/8e611d89cacc31c6ee3f2ae94cc34fc76402fd8e/tensorflow/compiler/xla/service/batchnorm_expander.cc) and [UnfuseBatchNormInference](https://github.com/tensorflow/mlir-hlo/blob/8ce6d6bea758ef9dac37ee67f2180d71b334d9fe/lib/Dialect/mhlo/transforms/unfuse_batch_norm.cc#L101). The HLO-based and MHLO-based passes perform computations a bit differently, perhaps accidentally or perhaps to improve numerical accuracy, but algebraically it's the same thing.

What I find somewhat lacking in Algorithm 1 in the spec, as far as speccing `batch_norm_inference` goes, is that it's not immediately clear how various inputs map on the notation in the paper. It's not super hard to figure out that gamma stands for scale, beta stands for offset, etc, but that's not immediately evident either. Also `feature_index` is something that doesn't come up in the algorithm at all.

Re: 2), I think it correlates very well with the inputs and explains how all of them are used, but it was pretty verbose. Also, MLIR notation isn't really designed for parameterization, so it wasn't clear how to generalize it from a single example with concrete shapes and concrete value of feature_index.

Instead, I propose that we use Python-based pseudocode to express the algorithm - I think the pseudocode has the strong sides of 1), 2) and 3) without having their downsides. Here's how the Semantics section could look like:

============
Normalizes the `operand` tensor across all dimensions except for the 
`feature_index` dimension and produces a `result` tensor.

```Python
def batch_norm_inference(operand, scale, offset, mean, variance, epsilon, feature_index):
  # Broadcast inputs to shape(operand)
  scale_bcast = broadcast_in_dim(scale, [feature_index], shape(operand))
  offset_bcast = broadcast_in_dim(offset, [feature_index], shape(operand))
  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
  variance_bcast = broadcast_in_dim(variance, [feature_index], shape(operand))
  epsilon_bcast = broadcast_in_dim(constant(epsilon), [], shape(operand))

  # Perform normalization using the provided `mean` and `variance` instead of 
  # computing them like `batch_norm_training` does.
  centered_operand = subtract(operand, mean_bcast)
  stddev = sqrt(add(variance_bcast, epsilon_bcast))
  normalized_operand = divide(centered_operand, stddev)
  return add(multiply(scale_bcast, normalized_operand), offset_bcast)
```
============